stone/ir/data_types.py

```diff
@@ -1645,6 +1711,45 @@ def __repr__(self):
         return 'RedactedHash(%r, %r, %r)' % (self.name, self.namespace, self.regex)
 
 
+class CustomAnnotation(Annotation):
+    """
+    Used when a field is annotated with a custom annotation type.
+    """
+    def __init__(self, name, namespace, ast_node, annotation_type,
+                 annotation_type_ns, args, kwargs):
+        super(CustomAnnotation, self).__init__(name, namespace, ast_node)
+        self.annotation_type_name = annotation_type
+        self.annotation_type_ns = annotation_type_ns
+        self.args = args
+        self.kwargs = kwargs
+
+        self.annotation_type = None
+
+    def set_attributes(self, annotation_type):
+        self.annotation_type = annotation_type
+
+        if self.args:
+            raise InvalidSpec('Annotations cannot accept positional arguments',
```

Comment:
There are two reasons why I left out support for positional arguments:

1. I was not sure if it was OK to assign semantic meaning to the ordering of the params (since, e.g., it seems that the ordering of struct fields is [not supposed to have a semantic meaning](https://github.com/dropbox/stone/blob/master/stone/ir/data_types.py#L736-L737))

2. It makes the logic for figuring out what value to assign to which parameter more complicated---you have to make sure that all positional arguments are supplied, that no argument is supplied both as positional and keyword, etc.

I think that this is a reasonable limitation for now. If we decide to actually reimplement built-in annotations as custom ones, it should be possible to add support for positional arguments as a backwards-compatible change.

Reply:
I think comment referred to in (1) is a bit misleading (some of the comments might be outdated). We do need to preserve the ordering of the struct fields, but not just for display only. Some backend languages may not support passing named arguments (eg. Obj-C). And the ordering of the struct fields would impact the order of the arguments passed to the routes.

We don't have to follow the rule of struct for annotation type. But it would be very confusing to have inconsistent Stone syntax for declaring built-in and custom annotations. See below example:
```
annotation InternalOnly = Omitted("internal")
annotation InternalOnly = Omitted(permission="internal") // would trigger compiler error
annotation ReallyNoteworthy = Noteworthy(importance="high")
annotation ReallyNoteworthy = Noteworthy("high") // would trigger compiler error
```

It's nice to have keyword arguments for annotation type. But positional arguments would be crucial for backward compatibility. We can refer to how Python deals with `args` and `kwargs` if we want to support both. Or if it's too complicated to do so for now, and we have to choose one, we should actually choose positional arguments.

Reply:
I went ahead and added support for positional arguments to custom annotations (and keyword arguments to builtin ones since it was easy, though I kept it undocumented). To keep things simple, I made it where you either specify all arguments positionally or all as keywords.

At the IR level, custom annotations will have both "args" and "kwargs" that are consistent with each other, so language backends can use whatever is easiest (even if the Stone spec uses a different form).

Does that approach resolve the concerns?